firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
+        else:
+            B_blocks = [B11.kron(B11)]
+    elif tdim == 3:
+        if formdegree == 0:
+            B_blocks = [kron3(B00, B00, B00)]
+        elif formdegree == 1:
+            B_blocks = [kron3(B00, B00, B11), kron3(B00, B11, B00), kron3(B11, B00, B00)]
+        elif formdegree == 2:
+            B_blocks = [kron3(B00, B11, B11), kron3(B11, B00, B11), kron3(B11, B11, B00)]
+        else:
+            B_blocks = [kron3(B11, B11, B11)]
+
+    B00.destroy()
+    B11.destroy()
+    if len(B_blocks) == 1:
+        result = B_blocks[0]
+    else:
+        nrows = sum(Bk.size[0] for Bk in B_blocks)
+        ncols = sum(Bk.size[1] for Bk in B_blocks)
+        csr_block = [Bk.getValuesCSR() for Bk in B_blocks]
+        ishift = numpy.cumsum([0] + [csr[0][-1] for csr in csr_block])
+        jshift = numpy.cumsum([0] + [Bk.size[1] for Bk in B_blocks])
+        indptr = numpy.concatenate([csr[0][bool(shift):]+shift for csr, shift in zip(csr_block, ishift[:-1])])
+        indices = numpy.concatenate([csr[1]+shift for csr, shift in zip(csr_block, jshift[:-1])])
+        data = numpy.concatenate([csr[2] for csr in csr_block])
+        result = PETSc.Mat().createAIJ((nrows, ncols), csr=(indptr, indices, data), comm=PETSc.COMM_SELF)
+        for B in B_blocks:
+            B.destroy()
+    return result
+
+
+def diff_matrix(tdim, formdegree, A00, A11, A10):
+    if formdegree == tdim:
+        ncols = A10.shape[0]**tdim
+        A_zero = PETSc.Mat().createAIJ((1, ncols), nnz=(0, 0), comm=PETSc.COMM_SELF)
+        A_zero.assemble()
+        return A_zero
+
+    A00 = petsc_sparse(A00)
+    A11 = petsc_sparse(A11)
+    A10 = petsc_sparse(A10)
+    if tdim == 1:
+        return A10
+    elif tdim == 2:
+        if formdegree == 0:
+            A_blocks = [[A00.kron(A10)], [A10.kron(A00)]]
+        elif formdegree == 1:
+            A_blocks = [[A10.kron(A11), A11.kron(A10)]]
+            A_blocks[-1][-1].scale(-1)
+    elif tdim == 3:
+        if formdegree == 0:
+            A_blocks = [[kron3(A00, A00, A10)], [kron3(A00, A10, A00)], [kron3(A10, A00, A00)]]
+        elif formdegree == 1:
+            size = tuple(A11.getSize()[k] * A10.getSize()[k] * A00.getSize()[k] for k in range(2))
+            A_zero = PETSc.Mat().createAIJ(size, nnz=(0, 0), comm=PETSc.COMM_SELF)
+            A_zero.assemble()
+            A_blocks = [[kron3(A00, A10, A11, scale=-1), kron3(A00, A11, A10), A_zero],
+                        [kron3(A10, A00, A11, scale=-1), A_zero, kron3(A11, A00, A10)],
+                        [A_zero, kron3(A10, A11, A00), kron3(A11, A10, A00, scale=-1)]]
+        elif formdegree == 2:
+            A_blocks = [[kron3(A10, A11, A11, scale=-1), kron3(A11, A10, A11), kron3(A11, A11, A10)]]
+
+    A00.destroy()
+    A11.destroy()
+    A10.destroy()
+    result = block_mat(A_blocks)
+    for A_row in A_blocks:
+        for A in A_row:
+            A.destroy()
+    return result
+
+
+def diff_prolongator(Vf, Vc, fbcs=[], cbcs=[]):
```

Comment:
Comment "Magic. Tabulate grad: H(grad) -> H(curl), or curl: H(curl) -> H(div), etc as an explicit sparse matrix. Works for any basis. These are the same matrices one needs for HypreAMS and friends."